compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -3239,3 +3240,48 @@ declare_lint! {
     Allow,
     "detects usage of old versions of or-patterns",
 }
+
+declare_lint! {
+    /// The `future_prelude_collision` lint detects the usage of trait methods which are ambiguous
+    /// with traits added to the prelude in future editions.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(future_prelude_collision)]
+    ///
+    /// trait Foo {
+    ///     fn try_into(self) -> Result<String, !>;
+    /// }
+    ///
+    /// impl Foo for &str {
+    ///     fn try_into(self) -> Result<String, !> {
+    ///         Ok(String::from(self))
+    ///     }
+    /// }
+    ///
+    /// fn main() {
+    ///     let x: String = "3".try_into().unwrap();
+    ///     //                  ^^^^^^^^
+    ///     // This call to try_into matches both Foo:try_into and TryInto::try_into as
+    ///     // `TryInto` has been added to the Rust prelude in 2021 edition.
+    ///     println!("{}", x);
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// In Rust 2021, one of the important introductions is the [prelude changes], which add
+    /// `TryFrom`, `TryInto`, and `FromIterator` into the standard library's prelude. Since this
+    /// results in an amiguity as to which method/function to call when an existing `try_into`
+    ///  method is called via dot-call syntax or a `try_from`/`from_iter` associated function
+    ///  is called directly on a type.
+    ///
+    /// [prelude changes]: https://blog.rust-lang.org/inside-rust/2021/03/04/planning-rust-2021.html#prelude-changes
+    pub FUTURE_PRELUDE_COLLISION,
+    Warn,
```

Comment:
As far as I looked, there is no such group yet. Also as for Allow instead of Warn--would you mind explaining the reasoning? Other future-compatibility/migration lints such as `BARE_TRAIT_OBJECTS` are Warn

Reply:
If you add this, I think it automatically ends up in the `rust-2021-compatibility` group:

```rust
    @future_incompatible = FutureIncompatibleInfo {
        reference: "issue #85684 <https://github.com/rust-lang/rust/issues/85684>",
        edition: Some(Edition::Edition2021),
    };
```

> Also as for Allow instead of Warn--would you mind explaining the reasoning?

We shouldn't cause warnings for people using Rust 2018 if they are not migrating to Rust 2021, unless it's something 'bad' or 'deprecated' that we want everyone in all editions to move away from.

E.g. KEYWORD_IDENTS is Allow. We don't warn about `async` as an identifier in Rust 2015, unless you ask specifically for migration warnings (e.g. `cargo fix --edition` or `-Wrust-2018-compatibility`).

Reply:
Makes sense, thanks for the explanation!